core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

```diff
@@ -598,13 +632,42 @@ private static void closeStatement(PreparedStatement preparedStatement) {
         if (preparedStatement != null) {
             try {
                 preparedStatement.close();
+            } catch (SQLRecoverableException ex) {
+                handleSQLRecoverableException(ex);
             } catch (SQLException e) {
                 log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
             }
         }
 
     }
 
+    private static void handleSQLRecoverableException(SQLRecoverableException ex){
+		Connection dbConnection;
+    	try {
+	    	if(dataSource != null){
+	    		dbConnection = getDBConnection(dataSource);
+				handleSQLRecoverableException(dbConnection, ex);
+	    	}
+    	} catch (SQLException e) {
+            log.error("Error occurred during SQLRecoverableException handling." + e.getMessage());
+    	}
+    }
+    
+    	private static void handleSQLRecoverableException(Connection dbConnection, SQLRecoverableException ex) {
```

Comment:
The connection is closing unexpectedly, without a clear indication as to why.    
When this event occurs the connection is left in an unknown state -- possibly abandoned.  
Rather than swallowing the exception, this attempts to close it cleanly.  
If left in the unknown state the connection may be unusable but still occupy a slot in the pool.  This connection may be served up to another request for resources.  If it isn't usable, it will again be returned to the pool and another request will be issued for a resource, This incurs another round trip which may, again get an unusable connection.  With the addition of each unusable connection in the pool, the performance will degrade.

Another change in this class exposes the logAbandoned property.  Implementing and enabling this property will indicate whether or not the connections are, in fact, unusable and the root cause will be indicated in the stack trace. 

Note: "Logging of abandoned Statements and Connections adds overhead for every Connection open or new Statement because a stack trace has to be generated."

Therefore, after addressing the root cause and eliminating the issue, disable this property.